tlair/src/main/scala/at/forsyte/apalache/tla/lir/transformations/standard/ReTLALanguagePred.scala

```diff
@@ -0,0 +1,179 @@
+package at.forsyte.apalache.tla.lir.transformations.standard
```

Comment:
It looks to me on first evaluation like pretty much all of the logic here is shared with, e.g., the [KeraLanguagePred](https://github.com/informalsystems/apalache/tree/unstable/tlair/src/main/scala/at/forsyte/apalache/tla/lir/transformations/standard/KeraLanguagePred.scala).

Would it make sense to try to abstract the core logic and share it between the different language preds?

IIUC, the core logic is mostly just concerned with traversing the AST. It almost looks to me like you could make `TlaEx` and instance of the [Foldable](https://eed3si9n.com/learning-scalaz/Foldable.html) type class, and then just use [`all`](https://javadoc.io/doc/org.scalaz/scalaz-core_2.12/latest/scalaz/Foldable.html) along with a predicate `isFooOper : TlaOper -> Boolean`, where `Foo` is defined for the particular instance of `LanguagePred`.

If something like this is feasible, I think you might be able to introduce this change while consolidating the logic of this family of classes, and maybe even yielding a net reduction in LOC. What do you think?

Reply:
It's a bit tricky, because that would _almost_ work. The problem with shared parentage is that 1) the difference is in one method's `case` classes, and that method is 95% of the class functionality (you could have a shared parent class with all the aux methods though) and 2) Unfortunately, reTLA isn't 100% a restriction on the allowed operators, because of functions. Since SANY expands `[f EXCEPT ![a] = b]` to `[f EXCEPT ![<<a>>] = b]` we have to support tuples, but only in function argument positions and not anywhere else. So you can't have `TlaFunOper.tuple -> false` or `TlaFunOper.tuple -> true`, because they're _sometimes_ allowed.

Reply:
I see. Thanks for considering.